sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowWriter.scala

```diff
@@ -224,6 +226,25 @@ private[arrow] class DoubleWriter(val valueVector: NullableFloat8Vector) extends
   }
 }
 
+private[arrow] class DecimalWriter(
+    val valueVector: NullableDecimalVector,
+    precision: Int,
+    scale: Int) extends ArrowFieldWriter {
+
+  override def valueMutator: NullableDecimalVector#Mutator = valueVector.getMutator()
+
+  override def setNull(): Unit = {
+    valueMutator.setNull(count)
+  }
+
+  override def setValue(input: SpecializedGetters, ordinal: Int): Unit = {
+    valueMutator.setIndexDefined(count)
+    val decimal = input.getDecimal(ordinal, precision, scale)
+    decimal.changePrecision(precision, scale)
+    DecimalUtility.writeBigDecimalToArrowBuf(decimal.toJavaBigDecimal, valueVector.getBuffer, count)
```

Comment:
This will be easier in Arrow 0.7. There is now APIs for directly setting values with `setSafe` for BigDecimal.

Reply:
@BryanCutler Thanks, I'll update it to use `setSafe` after upgrading Arrow to 0.7.

Btw, when I tested upgrading to 0.7 locally, `ArrowConvertersSuite.string type conversion` came to fail. Do you have any ideas of that?

Reply:
I haven't tried running those tests for 0.7 yet, let me see if I can reproduce

Reply:
I got the same error, let me dig into it a little

Reply:
It was an issue with StringWriter, I put the fix in #19284 please take a look, thanks!

Reply:
I've confirmed it fixes the failure. Thanks!